source/grid/grid_out.cc

```diff
@@ -3219,28 +3256,39 @@ namespace internal
                   << '\n';
             }
           else
-            // cell is at boundary and we
-            // are to treat curved
-            // boundaries. so loop over
-            // all faces and draw them as
-            // small pieces of lines
+            // cell is at boundary and we are to treat curved boundaries. so
+            // loop over all faces and draw them as small pieces of lines
             {
               for (unsigned int face_no=0;
                    face_no<GeometryInfo<dim>::faces_per_cell; ++face_no)
                 {
+                  point_cache.clear();
                   const typename dealii::Triangulation<dim,spacedim>::face_iterator
                   face = cell->face(face_no);
-                  if (face->at_boundary() || gnuplot_flags.curved_inner_cells)
+                  if (dim != spacedim || face->at_boundary() || gnuplot_flags.curved_inner_cells)
                     {
-                      // compute offset
-                      // of quadrature
-                      // points within
-                      // set of projected
-                      // points
+                      // compute offset of quadrature points within set of
+                      // projected points
                       const unsigned int offset=face_no*n_points;
-                      for (unsigned int i=0; i<n_points; ++i)
-                        out << (mapping->transform_unit_to_real_cell
-                                (cell, q_projector->point(offset+i)))
+                      point_cache.push_back(face->vertex(0));
+                      for (unsigned int i=1; i<n_points - 1; ++i)
+                        point_cache.push_back(mapping->transform_unit_to_real_cell
+                                              (cell, q_projector->point(offset+i)));
```

Comment:
I may not understand quite what the cache does, but this at least seems odd: you first push an untransformed point, and then several transformed points. Is this intended?

Reply:
Yes: we do not need to transform the Triangulation's vertices (its a little optimization). I will add a comment.